File: desktop/src/shared/apiHostRotation.ts
/**
 * Phase 10 hardening ("Пул резервных доменов для API сервера"): if the
 * primary API domain is blocked/poisoned, fall through to the next
 * configured backup domain rather than failing outright. Pure, in-memory,
 * per-process — no persistence of "which host worked last" across app
 * restarts. Same contract as the Android client's ApiHostRotation.
 */
export class ApiHostRotation {
  private readonly hosts: string[];
  private currentIndex = 0;

  constructor(hosts: string[]) {
    if (!hosts || hosts.length === 0) {
      throw new Error('At least one API host is required');
    }
    this.hosts = [...hosts];
  }

  current(): string {
    return this.hosts[this.currentIndex];
  }

  size(): number {
    return this.hosts.length;
  }

  /** Advances to the next host (wrapping around) and returns it. */
  advance(): string {
    this.currentIndex = (this.currentIndex + 1) % this.hosts.length;
    return this.current();
  }
}

File: android/app/src/main/java/com/vpn/android/api/ApiHostRotation.java
package com.vpn.android.api;

import java.util.ArrayList;
import java.util.List;

/**
 * Phase 10 hardening ("Пул резервных доменов для API сервера"): if the
 * primary API domain is blocked/poisoned, fall through to the next
 * configured backup domain rather than failing outright. Pure, in-memory,
 * per-process — no persistence of "which host worked last" across app
 * restarts, which is an acceptable simplification for this MVP pass.
 */
public class ApiHostRotation {

    private final List<String> hosts;
    private int currentIndex = 0;

    public ApiHostRotation(List<String> hosts) {
        if (hosts == null || hosts.isEmpty()) {
            throw new IllegalArgumentException("At least one API host is required");
        }
        this.hosts = new ArrayList<>(hosts);
    }

    public String current() {
        return hosts.get(currentIndex);
    }

    public int size() {
        return hosts.size();
    }

    /** Advances to the next host (wrapping around) and returns it. */
    public String advance() {
        currentIndex = (currentIndex + 1) % hosts.size();
        return current();
    }
}
